### PR TITLE
Added tasks to load warehouse data into BigQuery.

### PR DESCRIPTION
**Pull request by HassanJaveed84**
Question about the way I've defined the schema.
Is it okay to place technology specific code in record.py ? BigQuery schema definition is a bit different than our current mysql/hive/vertica stuff we have in record.py.

### CODE REVIEW

**Comment by azafty468** [comment]
So normally I wouldn't put technology specific functions into our generalized classes, but that level of generalized purity sometimes doesn't work in practice.  I'm fine with what you've done in the code here.

**Review by azafty468 — APPROVED** [review]
Looks good to me!  Do we know if we need to put a cool down or wait period after deleting a table (on the overwrite)?  Or is the delete immediately reflected in the API?

File: edx/analytics/tasks/common/bigquery_load.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
+import logging
+import datetime
+import os
+import luigi
+import json
+import time
+import subprocess
+import urlparse
+import tempfile
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py]: nit: imports not in order....

File: edx/analytics/tasks/common/bigquery_load.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            raise RuntimeError("Retry limit exceeded while waiting on job.")
+
+        job.reload()
+        if job.state == 'DONE':
+            if job.error_result:
+                raise RuntimeError(job.errors)
+            return
+
+        counter += 1
+        time.sleep(1)
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py]: What's the normal runtime of these jobs?  I ran with a sleep of 5, which seemed like adequate granularity.

File: edx/analytics/tasks/common/bigquery_load.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+def wait_for_job(job):
+    counter = 0
+    while True:
+        if counter == RETRY_LIMIT:
+            raise RuntimeError("Retry limit exceeded while waiting on job.")
+
+        job.reload()
+        if job.state == 'DONE':
+            if job.error_result:
+                raise RuntimeError(job.errors)
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py]: I found that I wanted more control over the handling of job failures, so it was easier for me to move these two lines to the code after the wait_for_job() was called.  We've still got the job object, so it's no different.  And then this really is *only* waiting for the job.

File: edx/analytics/tasks/common/bigquery_load.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            return
+
+        counter += 1
+        time.sleep(1)
+
+
+class BigQueryTarget(luigi.Target):
+
+    def __init__(self, credentials_target, dataset_id, table, update_id):
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py:45]: I have no idea, but is there a need for a super() call here?

File: edx/analytics/tasks/common/bigquery_load.py (hunk 3 above)
Comment (brianhw) [reply edx/analytics/tasks/common/bigquery_load.py]: Maybe that should just be an option, then.  check_error_result=True by default, and I can then disable it.

File: edx/analytics/tasks/common/bigquery_load.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+        source_path = self.input()['source'].path
+        parsed_url = urlparse.urlparse(source_path)
+        destination_path = url_path_join('gs://{}'.format(parsed_url.netloc), parsed_url.path)
+
+        if self.is_file(source_path):
+            return_code = subprocess.call(['gsutil', 'cp', source_path, destination_path])
+        else:
+            log.debug(" ".join(['gsutil', '-m', 'rsync', source_path, destination_path]))
+            return_code = subprocess.call(['gsutil', '-m', 'rsync', source_path, destination_path])
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py:191]: Should there be something here that checks the return_code to see if it failed or succeeded?  Presumably we don't want to call load_table_from_storage if the data is not in google storage.

File: edx/analytics/tasks/common/bigquery_load.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                update_id=self.update_id(),
+            )
+
+        return self.output_target
+
+    def update_id(self):
+        return '{task_name}(date={key})'.format(task_name=self.task_family, key=self.date.isoformat())
+
+    def is_file(self, path):
+        if path.endswith('.tsv') or path.endswith('.csv'):
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py]: This is adequate for this current use, but presumably it should also handle .gz as well?

File: edx/analytics/tasks/common/bigquery_load.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            load_uri
+        )
+        job.field_delimiter = self.field_delimiter
+        job.quote_character = self.quote_character
+        job.null_marker = self.null_marker
+
+        log.debug("Starting BigQuery Load job.")
+        job.begin()
+        wait_for_job(job)
+
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py]: Might be nice to have some statistics logged here.  Something like:
    try:
        log.debug("   Load job started: %s   ended: %s   input_files: %s  output_rows: %s  output_bytes: %s",
            job.started, job.ended, job.input_files, job.output_rows, job.output_bytes)
    except KeyError as keyerr:
        log.debug("   Load job started: %s   ended: %s   No load stats.", job.started, job.ended)

    if job.error_result:
        for error in job.errors:
            log.debug("   Load error: %s", error)
    else:
        log.debug("   No errors encountered!")

File: edx/analytics/tasks/common/bigquery_load.py (hunk 5 above)
Comment (brianhw) [reply edx/analytics/tasks/common/bigquery_load.py:191]: As I mentioned in Hipchat, I assume we can create a Google storage Client the same way we do for BigQuery. Then we just get a bucket, define a blob in it, and call upload_from_filename. Is what's missing the globbing?  Best I can tell, gsutil is required if we were streaming, and didn't have data in separate files.

**Review by brianhw — APPROVED** [review]
Looks good overall.  Mostly nits.  

Just a heads-up, however, that I'm going to have to figure out how to refactor some of this in order to work with events.  For them, we would be writing to partitions instead of tables.

File: edx/analytics/tasks/common/bigquery_load.py (hunk 1 above)
Comment (HassanJaveed84) [reply edx/analytics/tasks/common/bigquery_load.py]: To clarify, by order do you mean alphabetically ? I've always been confused on how we are supposed to order imports

File: edx/analytics/tasks/common/bigquery_load.py (hunk 2 above)
Comment (HassanJaveed84) [reply edx/analytics/tasks/common/bigquery_load.py]: A few seconds, let's default to 5.

File: edx/analytics/tasks/common/bigquery_load.py (hunk 4 above)
Comment (HassanJaveed84) [reply edx/analytics/tasks/common/bigquery_load.py:45]: No, luigi.Target does not define a constructor and is derived from 'object'.

File: edx/analytics/tasks/common/bigquery_load.py (hunk 5 above)
Comment (HassanJaveed84) [reply edx/analytics/tasks/common/bigquery_load.py:191]: Good catch, I'll add a check for return_code value.

File: edx/analytics/tasks/common/bigquery_load.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        destination_path = url_path_join('gs://{}'.format(parsed_url.netloc), parsed_url.path)
+
+        if self.is_file(source_path):
+            return_code = subprocess.call(['gsutil', 'cp', source_path, destination_path])
+        else:
+            log.debug(" ".join(['gsutil', '-m', 'rsync', source_path, destination_path]))
+            return_code = subprocess.call(['gsutil', '-m', 'rsync', source_path, destination_path])
+
+        if return_code != 0:
+            raise RuntimeError('Error while syncing {source} to {destination}'.foramt(
```
Comment (brianhw) [inline edx/analytics/tasks/common/bigquery_load.py]: foramt => format

**Pull request merged**